## Lesson 3. Pandas
[Software, Data and
Technology](https://lp.jetbrains.com/software-data-and-technology-constructor-university/) bachelor program at [Constructor
University](https://constructor.university) (Bremen).

*Ilya Schurov, 2024*

*The explanations are mostly auto-generated using Claude Sonnet-3.5 model*

### Basics of pandas
In this lesson, we'll explore Pandas, a powerful library for data manipulation and analysis in Python.

In [1]:
import pandas as pd
import numpy as np

We start by creating a Pandas Series, which is a one-dimensional labeled array. The `pd.Series()` function takes a list of values and an optional `index` parameter to specify custom labels. In this case, we create a Series with three values and assign them custom string labels 'a', 'b', and 'c'.

In [2]:
ser = pd.Series([0, 10, 20], index=["a", "b", "c"])

Pandas Series support vectorized operations, which means we can perform operations on the entire Series at once without using explicit loops. This feature makes data manipulation efficient and concise. In the following example, we multiply each element of the Series by 2:

In [3]:
ser * 2

a     0
b    20
c    40
dtype: int64

The result is a new Series with the same index as the original, but with each value doubled. Notice how the operation is applied element-wise without the need for explicit iteration. The `dtype` of the resulting Series is `int64`, which indicates 64-bit integer values. Pandas automatically preserves the index during operations, maintaining the relationship between labels and values.

pandas Series offer flexible indexing options. Let's explore label-based and position-based indexing:

In [4]:
ser["b"]

10

In this example, we use label-based indexing to access the value associated with the label 'b'. This is particularly useful when working with non-numeric indices.

Series can also be created directly from a dictionary, where keys become the index and values become the Series values:

In [5]:
pd.Series({"a": 0, "b": 10})

a     0
b    10
dtype: int64

This creates a pandas Series from a dictionary. The keys of the dictionary become the index of the Series, while the values become the data. The `dtype: int64` indicates that the data type of the Series is 64-bit integer.

For positional indexing, pandas provides the `iloc` accessor. This is useful when you need to access elements by their integer position:

In [6]:
ser.iloc[1]

10

Here, `ser.iloc[1]` retrieves the second element (index 1) of the Series. Remember, Python uses zero-based indexing, so `iloc[1]` refers to the second element.

Now, let's create two Series objects representing students' grades:

In [7]:
alice = pd.Series({"Algebra": 5, "Geometry": 3, "Music": 4})
bob = pd.Series({"Algebra": 4, "Music": 3, "History": 4})

These Series objects, `alice` and `bob`, represent their respective grades in different subjects. The keys are subject names, and the values are the corresponding grades. Notice that Alice and Bob have some overlapping subjects (Algebra and Music) and some unique subjects (Geometry for Alice, History for Bob). This setup is useful for demonstrating operations on Series with different indexes.

We can check the data type of the Series using the `dtype` attribute:

In [8]:
alice.dtype

dtype('int64')

The `dtype` of `int64` indicates that the values in the Series are 64-bit integers. This is important to know when performing operations or analysis on the data, as certain functions may require specific data types.

In [9]:
grade = alice + bob

We examine the result of adding two Series objects, `alice` and `bob`. Pandas aligns data based on the index (subject names) during operations. For subjects in both Series, grades are added. Subjects in only one Series result in NaN (Not a Number) due to missing values.

In [10]:
grade.dtype

dtype('float64')

The `dtype` of the resulting `grade` Series is `float64`, despite starting with `int64` Series. Pandas promotes the data type to handle potential NaN values from the addition. The `float64` type represents whole numbers, decimals, and special values like NaN.

In [11]:
grade

Algebra     9.0
Geometry    NaN
History     NaN
Music       7.0
dtype: float64

The `grade` Series displays the result of combining `alice` and `bob`. 'Algebra' and 'Music' show the sum of their grades, while 'Geometry' and 'History' show NaN (Not a Number) as they were present in only one of the original Series. This demonstrates how pandas handles missing data during operations, preserving all index labels and using NaN for absent values.

In [12]:
grade["Geometry"]

nan

In [13]:
float("NaN")

nan

In pandas, `NaN` (Not a Number) is used to represent missing or undefined values. It's important to note that `NaN` values behave differently from other numeric values and require special handling.

In [14]:
if grade["Geometry"] == float("NaN"):
    print("It's a NaN")
else:
    print("????!!!")

????!!!


The output above might be surprising. Despite both `grade['Geometry']` and `float("NaN")` appearing to be `NaN`, the comparison returns `False`. This is because `NaN` is not equal to itself in floating-point arithmetic, a behavior inherited from the IEEE 754 standard.

In [15]:
if float("NaN") == float("NaN"):
    print("It's a NaN")
else:
    print("????!!!")

????!!!


Okay, so checking equality does not work. Then, how we can determine whether some value is NaN? There exists special functions for that.

In [16]:
if pd.isna(float("NaN")):
    print("Yes")

Yes


The `pd.isna()` function is a reliable tool for identifying `NaN` (Not a Number) values in pandas. It returns `True` for `NaN`, which is particularly useful because `NaN` values behave unusually in comparisons - they're considered unequal to everything, including themselves.

In [17]:
pd.isna(None)

True

Interestingly, `pd.isna()` also returns `True` for `None` values. This is because pandas treats both `NaN` and `None` as representations of missing data, providing a consistent approach to handling missing values across different data types. This behavior is especially helpful when working with datasets that might use various ways to denote missing information.

It's worth comparing `pd.isna()` with NumPy's `np.isnan()` function to understand their differences:

In [18]:
np.isnan(float("NaN"))

True

In [19]:
np.isnan(None)

TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

This error demonstrates a limitation of NumPy's `isnan()` function when dealing with `None` values. NumPy's `isnan()` is designed primarily for numerical data and doesn't handle `None` values, which are often used to represent missing data in Python. This is where pandas' `isna()` function shows its versatility, as it can handle both `NaN` and `None` values seamlessly.

Let's examine a pandas Series object named `alice`:

In [20]:
alice

Algebra     5
Geometry    3
Music       4
dtype: int64

The `alice` Series represents a student's grades across different subjects. In this structure, the subject names (Algebra, Geometry, Music) serve as the index labels, while the corresponding integers are the grade values. The `dtype: int64` indicates that the values are stored as 64-bit integers.

We can access specific elements of the Series using various indexing methods:

In [21]:
alice["Algebra":"Geometry"]

Algebra     5
Geometry    3
dtype: int64

This demonstrates label-based indexing in pandas Series. We select a range of elements from 'Algebra' to 'Geometry', inclusive of both end points. This is different from Python's typical slicing behavior and is a feature specific to pandas.

In [23]:
alice[0:1]

Algebra    5
dtype: int64

When use positional indexing, the usual convention "last index is not included" is respected.

The `values` attribute of a pandas Series provides access to the underlying NumPy array. This can be useful for operations that are more efficient with NumPy or when interfacing with libraries that expect NumPy arrays.

In [24]:
alice.values

array([5, 3, 4])

By accessing `alice.values`, we retrieve the Series data as a NumPy array. This representation loses the index information but can be advantageous for certain numerical operations or when working with libraries that primarily use NumPy arrays.

In [76]:
df = pd.DataFrame(
    [["Alice", 1, 2, 3], ["Bob", 4, 5, 6], ["Claudia", 2, 3, 4]],
    columns=["Name", "Algebra", "Geometry", "Calculus"],
)

Here, we create a pandas DataFrame named `df`. The DataFrame is initialized with a list of lists, where each inner list represents a row of data. The `columns` parameter specifies the column names for the DataFrame. This structure allows us to represent tabular data with multiple columns, including both string and numeric data types.

In [77]:
df.dtypes

Name        object
Algebra      int64
Geometry     int64
Calculus     int64
dtype: object

The `dtypes` attribute of a DataFrame shows the data type of each column. Here, we see that the `Name` column is of type `object`, which is used for strings in pandas. The other columns (`Algebra`, `Geometry`, and `Calculus`) are of type `int64`, representing 64-bit integers. Understanding the data types is crucial for performing operations and ensuring data consistency in your analysis.

Why we use `object` instead of a proper string dtype in pandas? Let's investigate how numpy treats arrays with string dtypes.

In [78]:
x = np.array(["a", "b"])

In [79]:
x[0] = "hello, world"

Here, we attempt to assign a longer string to the first element of the NumPy array `x`. However, NumPy arrays have fixed-size elements, determined by the longest string at creation time. (Otherwise, it is impossible to make they so efficient.) If `x` was created with single-character strings, it can only store single characters.

In [80]:
x

array(['h', 'b'], dtype='<U1')

The output of `x` will show that only the first character `'h'` of `'hello, world'` was stored. This behavior highlights the limitations of NumPy arrays when working with variable-length string data. If we want to store variable-length strings (or other variable-length types of data) in an array, the only option is to store there not the data itself, but pointers to the corresponding objects that store the data. That's what pandas does by default with string text columns. That's why it uses `object` dtype for such columns. They are not processed as efficiently as fixed-length arrays, but when it comes to processing of text data it is usually more important to make sure that no data will be lost rather than optimize the processing time.

In [81]:
df

Name  Algebra  Geometry  Calculus
0    Alice        1         2         3
1      Bob        4         5         6
2  Claudia        2         3         4

The `df` variable is a pandas DataFrame containing student names and their scores in different subjects. The DataFrame has four columns: `Name`, `Algebra`, `Geometry`, and `Calculus`. Each row represents a student's data, allowing for easy manipulation and analysis of tabular data in Python.

In [82]:
df.set_index("Name")

Algebra  Geometry  Calculus
Name                                
Alice          1         2         3
Bob            4         5         6
Claudia        2         3         4

The `set_index()` method is used to change the index of the DataFrame. In this case, we set the `Name` column as the new index. This operation returns a new DataFrame where student names become the row labels, making it easier to access and manipulate data for specific students.

In [83]:
df.set_index("Name", inplace=True)

By using `inplace=True`, we modify the original `df` DataFrame directly. The `Name` column is removed from the DataFrame and becomes the index. This allows for more intuitive data access, as we can now use student names to select rows, like `df.loc['Alice']` to get Alice's scores.

In [84]:
df.index

Index(['Alice', 'Bob', 'Claudia'], dtype='object', name='Name')

The `df.index` command displays the current index of the DataFrame. After setting 'Name' as the index, we can see that the index now consists of student names. The output shows an Index object with the student names and additional metadata, such as the data type ('object' for strings) and the name of the index ('Name').

In [85]:
df.columns

Index(['Algebra', 'Geometry', 'Calculus'], dtype='object')

The `df.columns` attribute returns an Index object containing the column labels of the DataFrame. This is useful for verifying the structure of the DataFrame after modifications, showing us the remaining columns after 'Name' has been set as the index.

In [86]:
pd.DataFrame([[1, 20, 54], [13, 23, 54]])

0   1   2
0   1  20  54
1  13  23  54

This code demonstrates various ways to access data in a pandas DataFrame. The DataFrame `df` appears to contain information about students' scores in different subjects, with the students' names as the index.

In [87]:
df["Alice"]

KeyError: 'Alice'

The attempt to access `df['Alice']` results in a `KeyError`. This error occurs because `'Alice'` is part of the index (row labels), not a column name. To access data for Alice, we would use `df.loc['Alice']` instead.

In [88]:
df["Algebra"]

Name
Alice      1
Bob        4
Claudia    2
Name: Algebra, dtype: int64

Here, we successfully access the `'Algebra'` column using `df['Algebra']`. The output is a pandas Series object, which is a one-dimensional labeled array. The Series contains the Algebra scores for each student, with the student names as the index. This demonstrates how to retrieve a single column from a DataFrame.

In [89]:
df[0:2]

Algebra  Geometry  Calculus
Name                              
Alice        1         2         3
Bob          4         5         6

This code demonstrates integer-based indexing in pandas. `df[0:2]` selects the first two rows of the DataFrame. Note that unlike Python lists, the upper bound (2) is exclusive in pandas integer-based indexing. The output shows a subset of the original DataFrame with the specified rows and all columns. This method is useful for quickly viewing a portion of your data.

In [90]:
df.loc["Alice"]

Algebra     1
Geometry    2
Calculus    3
Name: Alice, dtype: int64

The `loc` accessor in pandas is a powerful tool for label-based indexing and selection. Let's explore its various uses:

In [91]:
df.loc["Alice", "Geometry"]

2

The `loc` accessor in pandas is used for label-based indexing. It allows us to select data from a DataFrame or Series based on the labels of rows and columns.

In [92]:
df.loc[:, "Geometry"]

Name
Alice      2
Bob        5
Claudia    3
Name: Geometry, dtype: int64

In this example, we use `loc` to select all rows (`:`) and the 'Geometry' column. This returns a Series containing Geometry scores for all students. The `:` acts as a wildcard, selecting all row labels.

In [93]:
df.loc[:, "Algebra":"Geometry"]

Algebra  Geometry
Name                      
Alice          1         2
Bob            4         5
Claudia        2         3

The `loc` accessor is used for label-based indexing, allowing selection of rows and columns by their labels. Here, we use it with a slice for both rows and columns. The `:` selects all rows, while `'Algebra':'Geometry'` selects a range of columns from 'Algebra' to 'Geometry' (inclusive). This returns a subset of the original DataFrame containing only the specified columns for all students.

In [94]:
df.iloc[0]

Algebra     1
Geometry    2
Calculus    3
Name: Alice, dtype: int64

In contrast to `loc`, the `iloc` accessor is used for integer-based indexing. `df.iloc[0]` selects the first row of the DataFrame (index 0) and returns it as a Series. This Series contains all the subject scores for Alice, the first student in the DataFrame. The output shows Alice's scores in each subject, with the index being the subject names and the values being the scores.

In [95]:
new_df = pd.DataFrame([[2, 3, 1], [5, 6, 7]], index=[1, 0])

This code demonstrates the creation and indexing of a pandas DataFrame with a custom index.

In [96]:
new_df

0  1  2
1  2  3  1
0  5  6  7

In [97]:
new_df.loc[0]

0    5
1    6
2    7
Name: 0, dtype: int64

In [98]:
new_df.iloc[0]

0    2
1    3
2    1
Name: 1, dtype: int64

When working with pandas DataFrames, we can access specific columns or subsets of data in various ways. Here, we're focusing on column and row selection techniques.

In [99]:
df[:2]["Algebra"]

Name
Alice    1
Bob      4
Name: Algebra, dtype: int64

In this example, we're selecting the 'Algebra' column for the first two rows of the DataFrame `df`. The result is a pandas Series containing the 'Algebra' scores for Alice and Bob. The `[:2]` slice selects the first two rows, and `['Algebra']` specifies the column we want to retrieve.

In [100]:
df.loc["Alice", "Algebra"] = 4

Here, we're attempting to update Alice's 'Algebra' score using chained indexing. However, this method is not recommended and can lead to unexpected behavior.

In [101]:
df

Algebra  Geometry  Calculus
Name                                
Alice          4         2         3
Bob            4         5         6
Claudia        2         3         4

The DataFrame `df` is displayed, showing the current state of our data. Note that Alice's 'Algebra' score is 4 at this point.

In [102]:
df[:2]["Algebra"][b"Alice"] = 2

/var/folders/h2/9nyrt4p55kq6pdvqg02_zmj40000gn/T/ipykernel_44519/1134890167.py:1: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df[:2]["Algebra"][b"Alice"] = 2
/var/folders/h2/9nyrt4p55kq6pdvqg02_zmj40000gn/T/ipykernel_44519/1134890167.py:1:

This section demonstrates the complexities of chained indexing and data views in pandas.

In [103]:
df

Algebra  Geometry  Calculus
Name                                
Alice          4         2         3
Bob            4         5         6
Claudia        2         3         4

The DataFrame `df` shows the updated scores after the previous operation. Alice's 'Algebra' score is 4, reflecting that chain assignment did not make changes to the initial DataFrame. In other circumstances, it can make changes, depending on whether the indexing operation returned view or copy.

In [104]:
short_df_algebra = df[:2]["Algebra"]

Here, we create a new Series `short_df_algebra` containing 'Algebra' scores for the first two students. This operation creates a view of the original DataFrame, not a copy. The relationship between `short_df_algebra` and `df` is complex and can lead to unexpected behavior in data manipulation.

In [105]:
short_df_algebra["Alice"] = 3

/var/folders/h2/9nyrt4p55kq6pdvqg02_zmj40000gn/T/ipykernel_44519/1938316182.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  short_df_algebra["Alice"] = 3


This operation attempts to modify Alice's 'Algebra' score in `short_df_algebra`. The warning message indicates that this approach might not reliably update the original DataFrame. Pandas issues this `SettingWithCopyWarning` to alert us about potential unexpected behavior. This scenario highlights the importance of using explicit data access methods like `loc` when modifying data in pandas.

In [106]:
short_df_algebra_copy = df[:2]["Algebra"].copy()

Here, we create `short_df_algebra_copy` using the `.copy()` method. This generates an actual copy of the data, not just a view. By working with a copy, we ensure that any modifications to `short_df_algebra_copy` won't affect the original DataFrame `df`. This approach prevents potential SettingWithCopyWarning issues and provides a more reliable way to manipulate subsets of data.

In [107]:
short_df_algebra_copy["Alice"] = 4

In [108]:
df.mean()

Algebra     3.000000
Geometry    3.333333
Calculus    4.333333
dtype: float64

The `df.mean()` function calculates the mean of each column in the DataFrame. It computes the average score for each subject across all students. The result is a Series with subject names as the index and their respective mean scores as values.

In [109]:
df.mean(axis=1)

Name
Alice      2.666667
Bob        5.000000
Claudia    3.000000
dtype: float64

Using `df.mean(axis=1)`, we calculate the mean along the rows (axis=1) instead of columns. This gives us the average score for each student across all subjects. The resulting Series has student names as the index and their overall average scores as values.

Now, let's add a new column to our DataFrame that represents the mean score for each student across all subjects:

In [110]:
df["mean"] = df.mean(axis=1)

The `mean()` function is applied with `axis=1`, which calculates the mean across columns (i.e., for each row). This new 'mean' column is added directly to our existing DataFrame `df`.

Next, we'll explore the `assign()` method, which allows us to create a new DataFrame with an additional column without modifying the original DataFrame:

In [111]:
df.assign(min=df.min(axis=1))

Algebra  Geometry  Calculus      mean  min
Name                                               
Alice          3         2         3  2.666667  2.0
Bob            4         5         6  5.000000  4.0
Claudia        2         3         4  3.000000  2.0

The `assign()` method creates a new DataFrame with an additional 'min' column. This column contains the minimum score for each student across all subjects, calculated using `df.min(axis=1)`. The `axis=1` parameter tells pandas to compute the minimum across columns for each row.

It's important to note that `assign()` returns a new DataFrame without modifying the original `df`. This is useful when you want to create temporary views of your data without altering the source DataFrame.

In [112]:
df

Algebra  Geometry  Calculus      mean
Name                                          
Alice          3         2         3  2.666667
Bob            4         5         6  5.000000
Claudia        2         3         4  3.000000

The DataFrame `df` currently contains columns for 'Algebra', 'Geometry', 'Calculus', and 'mean'. The 'mean' column represents the average score for each student across all subjects.

It's worth noting that when we use methods like `df.mean(axis=1)`, we're performing operations across rows (axis=1). This calculates the mean for each row, which in our case represents the average score for each student.

In [113]:
df

Algebra  Geometry  Calculus      mean
Name                                          
Alice          3         2         3  2.666667
Bob            4         5         6  5.000000
Claudia        2         3         4  3.000000

Now, let's add a new column `mean` to our DataFrame using direct assignment. This operation will modify the original DataFrame in-place.

The line `df['mean'] = df.mean(axis=1)` creates a new column named 'mean' and assigns it the row-wise mean values. This calculates each student's average score across all subjects.

In [114]:
df["mean"] = df.mean(axis=1)

After adding the 'mean' column, let's view the updated DataFrame:

In [115]:
df

Algebra  Geometry  Calculus      mean
Name                                          
Alice          3         2         3  2.666667
Bob            4         5         6  5.000000
Claudia        2         3         4  3.000000

The `mean` column has been added to the DataFrame, showing the average score for each student across all subjects. This direct assignment method modifies the DataFrame in-place.

Now, let's add another column `sum` to calculate the total score for each student:

In [116]:
df["sum"] = df.sum(axis=1)

This code creates a new 'sum' column by summing up the values across each row (axis=1). The `sum()` function with `axis=1` operates horizontally across columns for each row, calculating the total score for each student across all subjects, including the 'mean' column.

In [117]:
df

Algebra  Geometry  Calculus      mean        sum
Name                                                     
Alice          3         2         3  2.666667  10.666667
Bob            4         5         6  5.000000  20.000000
Claudia        2         3         4  3.000000  12.000000

The DataFrame now includes both `mean` and `sum` columns. The `sum` column values are higher than expected because they include the `mean` in the calculation. To get the sum of only the subject scores, we would need to sum only the specific subject columns.

Next, we add a 'min' column to show the lowest score for each student:

In [118]:
df_with_min = df.assign(min=df.min(axis=1))

The `assign()` method creates a new DataFrame `df_with_min` with an additional 'min' column. The `min()` function with `axis=1` finds the minimum value across each row. This new DataFrame includes all the original columns plus the new 'min' column, without modifying the original `df`.

Let's examine the new DataFrame:

In [119]:
df_with_min

Algebra  Geometry  Calculus      mean        sum  min
Name                                                          
Alice          3         2         3  2.666667  10.666667  2.0
Bob            4         5         6  5.000000  20.000000  4.0
Claudia        2         3         4  3.000000  12.000000  2.0

The `min` column displays the lowest score for each student across all subjects. This can be useful for quickly identifying a student's weakest subject or comparing performance across different students.

Now, let's compare this with the original DataFrame:

In [121]:
df

Algebra  Geometry  Calculus      mean        sum
Name                                                     
Alice          3         2         3  2.666667  10.666667
Bob            4         5         6  5.000000  20.000000
Claudia        2         3         4  3.000000  12.000000

The original DataFrame remains unchanged, demonstrating that `assign()` creates a new DataFrame without modifying the existing one. This non-destructive approach is particularly useful when you want to preserve the original data while creating new derived columns for analysis.

Now, let's filter the DataFrame to show only students with an Algebra score less than 3:

In [122]:
df[df["Algebra"] < 3]

Algebra  Geometry  Calculus  mean   sum
Name                                            
Claudia        2         3         4   3.0  12.0

Boolean indexing is a powerful feature in pandas that allows for efficient filtering of data. The condition `df['Algebra'] < 3` creates a boolean mask, which is then applied to the DataFrame to select only the rows where the condition is True.

Let's examine the boolean mask created by this condition:

In [123]:
df["Algebra"] < 3

Name
Alice      False
Bob        False
Claudia     True
Name: Algebra, dtype: bool

The `assign()` method in pandas allows us to create new columns in a DataFrame. It can be combined with filtering to create new columns and then filter the results. In the following example, we create a new column 'min' and then filter for students with an Algebra score less than 4:

In [124]:
df.assign(min=df.min(axis=1))[lambda x: x["Algebra"] < 4]

Algebra  Geometry  Calculus      mean        sum  min
Name                                                          
Alice          3         2         3  2.666667  10.666667  2.0
Claudia        2         3         4  3.000000  12.000000  2.0

### Case study: IMDb-5000 dataset

In this section, we'll be working with a new dataset containing movie metadata. We use the `pd.read_csv()` function to load the data from a URL directly into a pandas DataFrame called `df`.

The dataset is sourced from GitHub and contains information about various movies. Let's examine its contents:

In [125]:
df = pd.read_csv(
    "https://github.com/anishshah23/IMDb-5000-Data-analysis/raw/master/movie_metadata.csv"
)

After loading the data, it's a good practice to inspect the DataFrame to understand its structure and contents. We can use methods like `head()`, `info()`, or `describe()` to get an overview of the data.

In [126]:
df.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

The `head()` method displays the first few rows of the DataFrame, giving us a glimpse of the movie metadata. This allows us to see the column names and the type of information stored in each column.

Let's check the basic information about the DataFrame:

In [127]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5024 non-null   object 
 1   director_name              4939 non-null   object 
 2   num_critic_for_reviews     4993 non-null   float64
 3   duration                   5028 non-null   float64
 4   director_facebook_likes    4939 non-null   float64
 5   actor_3_facebook_likes     5020 non-null   float64
 6   actor_2_name               5030 non-null   object 
 7   actor_1_facebook_likes     5036 non-null   float64
 8   gross                      4159 non-null   float64
 9   genres                     5043 non-null   object 
 10  actor_1_name               5036 non-null   object 
 11  movie_title                5043 non-null   object 
 12  num_voted_users            5043 non-null   int64  
 13  cast_total_facebook_likes  5043 non-null   int64

The `info()` method provides a concise summary of the DataFrame, including the number of non-null values in each column and their data types. This is useful for identifying potential missing data or unexpected data types.

To get a more detailed view of the data types for each column, we can use the `dtypes` attribute:

In [128]:
df.dtypes

color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
movie_title                   object
num_voted_users                int64
cast_total_facebook_likes      int64
actor_3_name                  object
facenumber_in_poster         float64
plot_keywords                 object
movie_imdb_link               object
num_user_for_reviews         float64
language                      object
country                       object
content_rating                object
budget                       float64
title_year                   float64
actor_2_facebook_likes       float64
imdb_score                   float64
aspect_ratio                 float64
m

The `dtypes` attribute shows the data type of each column in the DataFrame. This is particularly useful for understanding how pandas has interpreted our data. We can see that:

- Categorical data like `color`, `director_name`, and `genres` are stored as `object` data type, which is pandas' way of handling string data.
- Numerical data like `num_critic_for_reviews`, `duration`, and `gross` are stored as `float64`, which allows for decimal values.
- Some integer data like `num_voted_users` and `cast_total_facebook_likes` are stored as `int64`.

Understanding these data types is crucial for further data manipulation and analysis, as certain operations may require specific data types. For example, statistical operations typically work with numerical data types, while text processing functions work with object (string) data types.

Let's take a closer look at the detailed output of `info()`:

In [129]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5024 non-null   object 
 1   director_name              4939 non-null   object 
 2   num_critic_for_reviews     4993 non-null   float64
 3   duration                   5028 non-null   float64
 4   director_facebook_likes    4939 non-null   float64
 5   actor_3_facebook_likes     5020 non-null   float64
 6   actor_2_name               5030 non-null   object 
 7   actor_1_facebook_likes     5036 non-null   float64
 8   gross                      4159 non-null   float64
 9   genres                     5043 non-null   object 
 10  actor_1_name               5036 non-null   object 
 11  movie_title                5043 non-null   object 
 12  num_voted_users            5043 non-null   int64  
 13  cast_total_facebook_likes  5043 non-null   int64

The `info()` method provides a concise summary of the DataFrame, including the number of non-null values and data types for each column. This is particularly useful for identifying missing data and understanding the structure of our dataset.

We can see that we have 5043 entries (rows) and 28 columns. The 'Non-Null Count' shows how many non-null values exist for each feature, helping us identify columns with missing data. For example, `budget` has 4551 non-null values, indicating that some movies lack budget information.

This information is crucial for data preprocessing. We might need to handle missing values or consider converting some `float64` columns to `int64` if they represent whole numbers (e.g., `title_year`).

In [130]:
df

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
5038  Color        Scott Smith                     1.0      87.0   
5039  Color                NaN                    43.0      43.0   
5040  Color   Benjamin Roberds                    13.0      76.0   
5041  Color        Daniel Hsia                    14.0     100.0   
5042  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
5038                      2.0                   318.0     Daphne Zuniga   
5039                      NaN                   319.0     Valorie Curry   
5040                      0.0                     0.0     Maxwell Moody   
5041                      0.0                   489.0     Daniel Henney   
5042                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
5038                   637.0          NaN                     Comedy|Drama   
5039                   841.0          NaN     Crime|Drama|Mystery|Thriller   
5040                     0.0          NaN            Drama|Horror|Thriller   
5041                   946.0      10443.0             Comedy|Drama|Romance   
5042                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
5038  ...                  6.0  English   Canada             NaN          NaN   
5039  ...                359.0  English      USA           TV-14          NaN   
5040  ...                  3.0  English      USA             NaN       1400.0   
5041  ...                  9.0  English      USA           PG-13          NaN   
5042  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

The output of `df` displays the entire DataFrame, giving us a comprehensive view of our movie dataset. Each row represents a unique movie, and each column represents a specific attribute of the movie.

The `genres` column contains multiple genres separated by '|', indicating that movies can belong to multiple categories. This might require special handling if we want to analyze movies by genre, such as splitting the genres into separate columns or using text processing techniques.

The `imdb_score` column provides the IMDb rating for each movie, which could be a key metric for analysis. We might want to explore the relationship between this score and other features like budget, director, or release year.

In [131]:
df["color"].value_counts(dropna=False)

color
Color               4815
 Black and White     209
NaN                   19
Name: count, dtype: int64

Let's analyze the distribution of movie colors in our dataset using the `value_counts()` method.

In [132]:
print(df["color"].value_counts(dropna=False))

color
Color               4815
 Black and White     209
NaN                   19
Name: count, dtype: int64


The `value_counts()` method counts the occurrences of each unique value in the `color` column. The `dropna=False` parameter ensures that NaN (missing) values are included in the count.

From the output, we can observe:
- The majority of movies (4815) are in color
- 209 movies are in black and white
- 19 movies have missing color information (NaN)

This distribution gives us insight into the prevalence of color in our movie dataset and identifies the presence of missing values. We might consider how to handle these missing values in our subsequent analysis.

In [133]:
df.query('color == "Black and White"')

Empty DataFrame
Columns: [color, director_name, num_critic_for_reviews, duration, director_facebook_likes, actor_3_facebook_likes, actor_2_name, actor_1_facebook_likes, gross, genres, actor_1_name, movie_title, num_voted_users, cast_total_facebook_likes, actor_3_name, facenumber_in_poster, plot_keywords, movie_imdb_link, num_user_for_reviews, language, country, content_rating, budget, title_year, actor_2_facebook_likes, imdb_score, aspect_ratio, movie_facebook_likes]
Index: []

[0 rows x 28 columns]

We encountered an issue when trying to filter the DataFrame for "Black and White" movies using the `query()` method. Let's investigate further.

In [134]:
df[df["color"] == " Black and White"]

color      director_name  num_critic_for_reviews  duration  \
111    Black and White        Michael Bay                   191.0     184.0   
149    Black and White       Lee Tamahori                   264.0     133.0   
257    Black and White    Martin Scorsese                   267.0     170.0   
272    Black and White       Michael Mann                   174.0     165.0   
286    Black and White    Martin Campbell                   400.0     144.0   
...                ...                ...                     ...       ...   
5005   Black and White    Andrew Bujalski                    52.0     109.0   
5008   Black and White        Kevin Smith                   136.0     102.0   
5015   Black and White  Richard Linklater                    61.0     100.0   
5022   Black and White         Jim Chuchu                     6.0      60.0   
5028   Black and White      Ivan Kavanagh                    12.0      83.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
111                       0.0                   691.0          Jaime King   
149                      93.0                   746.0        Colin Salmon   
257                   17000.0                   827.0          Adam Scott   
272                       0.0                   780.0  Jada Pinkett Smith   
286                     258.0                   834.0      Tobias Menzies   
...                       ...                     ...                 ...   
5005                     26.0                     3.0    Kate Dollenmayer   
5008                      0.0                   216.0    Brian O'Halloran   
5015                      0.0                     0.0   Richard Linklater   
5022                      0.0                     4.0       Olwenya Maina   
5028                     18.0                     0.0       Michael Parle   

      actor_1_facebook_likes        gross                            genres  \
111                   3000.0  198539855.0  Action|Drama|History|Romance|War   
149                    769.0  160201106.0         Action|Adventure|Thriller   
257                  29000.0  102608827.0                   Biography|Drama   
272                  10000.0   58183966.0             Biography|Drama|Sport   
286                   6000.0  167007184.0         Action|Adventure|Thriller   
...                      ...          ...                               ...   
5005                    26.0          NaN                            Comedy   
5008                   898.0    3151130.0                            Comedy   
5015                     5.0    1227508.0                      Comedy|Drama   
5022                   147.0          NaN                             Drama   
5028                    10.0          NaN                            Horror   

      ... num_user_for_reviews language  country  content_rating       budget  \
111   ...               1999.0  English      USA           PG-13  140000000.0   
149   ...               1185.0  English       UK           PG-13  142000000.0   
257   ...                799.0  English      USA           PG-13  110000000.0   
272   ...                386.0  English      USA               R  107000000.0   
286   ...               2301.0  English       UK           PG-13  150000000.0   
...   ...                  ...      ...      ...             ...          ...   
5005  ...                 23.0  English      USA               R          NaN   
5008  ...                615.0  English      USA               R     230000.0   
5015  ...                 80.0  English      USA               R      23000.0   
5022  ...                  1.0  Swahili    Kenya             NaN      15000.0   
5028  ...                  1.0  English  Ireland             NaN      10000.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
111       2001.0                  961.0        6.1          2.35   
149       2002.0                  766.0        6.1          2.35   
257       2004.0 

The previous code utilized boolean indexing to filter the DataFrame, revealing an unexpected space before "Black and White" in the condition. This successfully returned a DataFrame with 209 rows, matching our earlier count.

This discovery highlights the importance of data cleaning and consistency in string values when working with real-world datasets. Let's investigate further and clean our data.

In [135]:
print(df["color"].str.strip().value_counts())

color
Color              4815
Black and White     209
Name: count, dtype: int64


To ensure data consistency and accuracy, we'll clean the `color` column by removing any leading or trailing whitespace. This is an important step in data preprocessing.

In [136]:
df["color"] = df["color"].str.strip()

We start by examining the value counts of the `color` column to understand its distribution.

In [137]:
print(df["color"].value_counts())

color
Color              4815
Black and White     209
Name: count, dtype: int64


We'll now analyze the duration of movies based on their color classification.

In [138]:
df = pd.read_csv(
    "https://github.com/anishshah23/IMDb-5000-Data-analysis/raw/master/movie_metadata.csv"
)
string_columns = df.select_dtypes(include="object").columns
for column in string_columns:
    df[column] = df[column].str.strip()

Now we'll analyze the relationship between movie color and duration.

In [139]:
df.query('color == "Color"')["duration"].mean()

107.04290772755675

This code calculates the average duration of black and white movies using the `query()` method to filter the DataFrame.

In [140]:
df.query('color == "Black and White"')["duration"].mean()

112.2535885167464

The output shows that black and white movies in this dataset have an average duration of approximately 112.25 minutes.

In [141]:
df.groupby("color", dropna=False)["duration"].mean()

color
Black and White    112.253589
Color              107.042908
NaN                 90.722222
Name: duration, dtype: float64

This code groups the movies by their color category and calculates the mean duration for each group. The `dropna=False` parameter ensures that movies with missing color information (NaN) are included in the results. 

Interestingly, black and white movies have the longest average duration (112.25 minutes), followed by color movies (107.04 minutes). Movies with missing color information have the shortest average duration (90.72 minutes).

In [142]:
np.array([3, 4, np.nan, 2]).mean()

nan

This example demonstrates how NumPy and pandas handle `NaN` values when calculating the mean. NumPy's `mean()` function returns `NaN` if the array contains any `NaN` values, while pandas' `Series.mean()` method automatically excludes `NaN` values by default.

In [143]:
np.nanmean(np.array([3, 4, np.nan, 2]))

3.0

The `np.nanmean()` function is used to calculate the mean while ignoring `NaN` values in NumPy arrays. This function is particularly useful when working with datasets that may contain missing values, as it provides a way to compute the mean without manually filtering out `NaN` values.

In [144]:
pd.Series([3, 4, np.nan, 2]).mean()

3.0

As we can see, pandas' `Series.mean()` method automatically ignores `NaN` values, producing the same result as `np.nanmean()`. This behavior simplifies calculations on data with missing values, as it doesn't require explicit handling of `NaN` values for every operation.

In [145]:
pd.Series([3, 4, np.nan, 2]).fillna(0).mean()

2.25

Here, we use the `fillna()` method to replace `NaN` values with 0 before calculating the mean. This approach results in a different outcome compared to ignoring `NaN` values, as it treats missing data as actual zero values. The choice between ignoring `NaN` values or replacing them depends on the specific requirements of the data analysis task at hand.

In [146]:
df["director_name"].value_counts()["Tim Burton"]

16

This code snippet explores the distribution of directors in our dataset using the `value_counts()` method from pandas.

In [147]:
df.query('director_name == "Tim Burton"').shape

(16, 28)

Here, we use the `query()` method to filter the DataFrame for movies directed by Tim Burton. The `shape` attribute returns a tuple representing the dimensions of the resulting DataFrame. The output (16, 28) indicates that there are 16 movies by Tim Burton in our dataset, with 28 columns of information for each movie.

In [148]:
df["director_name"].value_counts().sort_values(ascending=False)

director_name
Steven Spielberg    26
Woody Allen         22
Clint Eastwood      20
Martin Scorsese     20
Ridley Scott        17
                    ..
Ryan Smith           1
Travis Romero        1
Andrew Haigh         1
Cary Bell            1
Daniel Hsia          1
Name: count, Length: 2398, dtype: int64

This code analyzes the frequency of directors in the dataset. The `value_counts()` method counts the occurrences of each director, and `sort_values(ascending=False)` arranges the results in descending order. Steven Spielberg leads with 26 movies, followed by Woody Allen with 22. The output reveals that there are 2398 unique directors in the dataset, with many having directed only one movie.

In [149]:
df["director_name"].value_counts().sort_index()

director_name
A. Raven Cruz         1
Aaron Hann            1
Aaron Schneider       1
Aaron Seltzer         1
Abel Ferrara          1
                     ..
Zoran Lisinac         1
Álex de la Iglesia    1
Émile Gaudreault      1
Éric Tessier          1
Étienne Faure         1
Name: count, Length: 2398, dtype: int64

The following code snippets demonstrate various operations on the director data in the DataFrame.

In [150]:
df["director_name"].str.split(expand=True)

0          1     2     3
0           James    Cameron  None  None
1            Gore  Verbinski  None  None
2             Sam     Mendes  None  None
3     Christopher      Nolan  None  None
4            Doug     Walker  None  None
...           ...        ...   ...   ...
5038        Scott      Smith  None  None
5039          NaN        NaN   NaN   NaN
5040     Benjamin    Roberds  None  None
5041       Daniel       Hsia  None  None
5042          Jon       Gunn  None  None

[5043 rows x 4 columns]

The following code demonstrates how to extract and analyze director names from the dataset.

In [151]:
df_director_counts = df["director_name"].value_counts().to_frame().reset_index()
df_director_counts = (
    pd.concat(
        [
            df_director_counts,
            df_director_counts["director_name"].str.split(expand=True).iloc[:, 1],
        ],
        axis=1,
    )
    .rename(columns={1: "director_last_name"})
    .sort_values("director_last_name")
)

This code manipulates the `df_director_counts` DataFrame, which contains information about directors and their movie counts. Let's analyze the operations and their results:

In [152]:
df_director_counts

director_name  count director_last_name
1239      John 'Bud' Cardos      1              'Bud'
1727         Brian A Miller      1                  A
1790      William A. Fraker      1                 A.
42         George A. Romero      9                 A.
1282  Marius A. Markevicius      1                 A.
...                     ...    ...                ...
1374              Valentine      1               None
1469                   Remo      1               None
1868                  Pitof      1               None
2089                Maïwenn      1               None
2276                    RZA      1               None

[2398 rows x 3 columns]

This section explores the dataset further, focusing on director counts and average IMDb scores by country.

In [153]:
df_director_counts["director_name"].str.split(expand=True).iloc[:, 1]

1239    'Bud'
1727        A
1790       A.
42         A.
1282       A.
        ...  
1374     None
1469     None
1868     None
2089     None
2276     None
Name: 1, Length: 2398, dtype: object

The `df_director_counts` DataFrame shows the count of movies for each director in the dataset. It includes the director's full name, the count of their movies, and what appears to be their last name. The presence of `NaN` values indicates potential data quality issues that might need addressing in future analysis.

In [154]:
df_director_counts

director_name  count director_last_name
1239      John 'Bud' Cardos      1              'Bud'
1727         Brian A Miller      1                  A
1790      William A. Fraker      1                 A.
42         George A. Romero      9                 A.
1282  Marius A. Markevicius      1                 A.
...                     ...    ...                ...
1374              Valentine      1               None
1469                   Remo      1               None
1868                  Pitof      1               None
2089                Maïwenn      1               None
2276                    RZA      1               None

[2398 rows x 3 columns]

This code calculates the average IMDb score for movies from each country, sorted in descending order. The results provide insights into the relationship between a movie's country of origin and its IMDb rating.

In [155]:
df.groupby("country")["imdb_score"].mean().sort_values(ascending=False)

country
Kyrgyzstan              8.7
Libya                   8.4
United Arab Emirates    8.2
Egypt                   8.1
Soviet Union            8.1
                       ... 
Georgia                 5.6
Peru                    5.4
Aruba                   4.8
Bahamas                 4.4
New Line                4.4
Name: imdb_score, Length: 65, dtype: float64

The previous analysis didn't account for the number of movies from each country, which could lead to misleading results due to countries with very few films. To address this, we can include the count of movies along with the mean score and filter for countries with a significant number of films.

In [156]:
(
    df.groupby("country")["imdb_score"]
    .agg(["mean", "count"])
    .sort_values("mean", ascending=False)
    .query("count > 20")
)

mean  count
country                   
Japan      6.952174     23
Italy      6.873913     23
Spain      6.824242     33
UK         6.818304    448
France     6.678571    154
China      6.623333     30
India      6.532353     34
Australia  6.514545     55
USA        6.367428   3807
Germany    6.340206     97
Canada     6.161905    126

Now that we have an overview of the dataset, let's identify each director's best movie based on IMDb scores. This analysis will provide insights into which directors consistently produce highly-rated films and what their best works are according to IMDb ratings.

Now, let's solve the following problem:

*for each director, find his or her best movie*

In [158]:
director_best_scores = (
    df.groupby("director_name")["imdb_score"].max().to_frame().reset_index()
)

This code snippet creates a new dataframe `director_best_scores` that contains each director's highest IMDb score. The `groupby('director_name')` function groups the original dataframe by director, and `['imdb_score'].max()` selects the highest score for each group. The `to_frame()` method converts the result to a dataframe, and `reset_index()` ensures that 'director_name' becomes a regular column instead of an index.

In [159]:
df

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
5038  Color        Scott Smith                     1.0      87.0   
5039  Color                NaN                    43.0      43.0   
5040  Color   Benjamin Roberds                    13.0      76.0   
5041  Color        Daniel Hsia                    14.0     100.0   
5042  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
5038                      2.0                   318.0     Daphne Zuniga   
5039                      NaN                   319.0     Valorie Curry   
5040                      0.0                     0.0     Maxwell Moody   
5041                      0.0                   489.0     Daniel Henney   
5042                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
5038                   637.0          NaN                     Comedy|Drama   
5039                   841.0          NaN     Crime|Drama|Mystery|Thriller   
5040                     0.0          NaN            Drama|Horror|Thriller   
5041                   946.0      10443.0             Comedy|Drama|Romance   
5042                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
5038  ...                  6.0  English   Canada             NaN          NaN   
5039  ...                359.0  English      USA           TV-14          NaN   
5040  ...                  3.0  English      USA             NaN       1400.0   
5041  ...                  9.0  English      USA           PG-13          NaN   
5042  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

The `df` dataframe contains our original dataset with various movie attributes. We'll now explore this data further and combine it with the `director_best_scores` dataframe to gain insights into directors' best-rated films.

In [160]:
director_best_scores

director_name  imdb_score
0          A. Raven Cruz         1.9
1             Aaron Hann         6.0
2        Aaron Schneider         7.1
3          Aaron Seltzer         2.7
4           Abel Ferrara         6.6
...                  ...         ...
2393       Zoran Lisinac         7.1
2394  Álex de la Iglesia         6.1
2395    Émile Gaudreault         6.7
2396        Éric Tessier         6.6
2397       Étienne Faure         4.3

[2398 rows x 2 columns]

The `director_best_scores` dataframe shows each director's highest IMDb score. This condensed view allows for quick comparisons between directors based on their top-rated films.

In [161]:
df.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

Displaying `df.columns` gives us an overview of all available features in our dataset. Key columns include `director_name`, `imdb_score`, `gross`, `budget`, and `movie_title`, which are crucial for analyzing movie performance and reception.

In [162]:
df[["director_name", "movie_title", "title_year", "imdb_score"]].merge(
    director_best_scores.rename(columns={"imdb_score": "best_imdb_score"}),
    on="director_name",
).query("imdb_score == best_imdb_score").sort_values("imdb_score", ascending=False)

director_name               movie_title  title_year  imdb_score  \
2712        John Blanchard          Towering Inferno         NaN         9.5   
1902        Frank Darabont  The Shawshank Redemption      1994.0         9.3   
3401  Francis Ford Coppola             The Godfather      1972.0         9.2   
4315        John Stockwell      Kickboxer: Vengeance      2016.0         9.1   
66       Christopher Nolan           The Dark Knight      2008.0         9.0   
...                    ...                       ...         ...         ...   
4669        Georgia Hilton              Subconscious      2015.0         2.2   
3278    Vondie Curtis-Hall                   Glitter      2001.0         2.1   
1698     Frédéric Auburtin           United Passions      2014.0         2.0   
4507         A. Raven Cruz       The Helix... Loaded      2005.0         1.9   
1115     Lawrence Kasanoff                Foodfight!      2012.0         1.7   

      best_imdb_score  
2712              9.5  
1902              9.3  
3401              9.2  
4315              9.1  
66                9.0  
...               ...  
4669              2.2  
3278              2.1  
1698              2.0  
4507              1.9  
1115              1.7  

[2508 rows x 5 columns]

This operation combines data from `df` and `director_best_scores` to create a view of directors' best-rated movies. It selects relevant columns, merges the dataframes on the `director_name` column, and filters for movies where the `imdb_score` matches the `best_imdb_score`. The result is sorted by `imdb_score` in descending order, effectively showing each director's top-rated film. This allows us to compare the most critically acclaimed works across different directors.

In [163]:
left = pd.DataFrame({"x": ["a", "b"], "y": [1, 2]})
right = pd.DataFrame({"x": ["a", "w", "a"], "u": [1, 2, 3]})

Here, we create two sample DataFrames, `left` and `right`, to demonstrate different types of joins in pandas. The `left` DataFrame has columns 'x' and 'y', while the `right` DataFrame has columns 'x' and 'u'. The 'x' column is common to both DataFrames and will serve as the key for joining operations. Notice that `right` has a duplicate value 'a' in the 'x' column, which will be important when we explore various join types.

In [164]:
left

x  y
0  a  1
1  b  2

In [165]:
right

x  u
0  a  1
1  w  2
2  a  3

In [166]:
left.merge(right, on="x", how="left")

x  y    u
0  a  1  1.0
1  a  1  3.0
2  b  2  NaN

The `merge` function is used to combine DataFrames. In this example, we perform a left join using `left.merge(right, on='x', how='left')`. The `on='x'` parameter specifies that we're joining on the 'x' column, and `how='left'` indicates a left join. This means all rows from the `left` DataFrame are included, and matching rows from the `right` DataFrame are added. When there's no match in `right`, NaN values are filled in.

Notice how the result has two rows for 'a' because there were two matching rows in the `right` DataFrame. The 'b' row has NaN for 'u' because there was no matching 'b' in the `right` DataFrame. This demonstrates how left joins handle both one-to-many relationships and unmatched keys.

In [167]:
left.merge(right, on="x", how="right")

x    y  u
0  a  1.0  1
1  w  NaN  2
2  a  1.0  3

Now we perform a right join using `left.merge(right, on='x', how='right')`. This operation includes all rows from the `right` DataFrame and matching rows from the `left` DataFrame. Observe that the result includes all values from the `right` DataFrame, including the 'w' row which doesn't have a match in `left`. The 'y' column for this row contains NaN. Also, note that both 'a' rows from `right` are present, each matched with the single 'a' row from `left`. This illustrates how right joins handle unmatched keys and one-to-many relationships from the perspective of the right DataFrame.

In [168]:
left.merge(right, on="x", how="outer")

x    y    u
0  a  1.0  1.0
1  a  1.0  3.0
2  b  2.0  NaN
3  w  NaN  2.0

Here, we perform an outer join using `left.merge(right, on='x', how='outer')`. An outer join includes all rows from both DataFrames, filling in NaN values where there are no matches. This result combines all unique keys from both DataFrames. Notice how it includes the 'b' row from `left` (with NaN for 'u') and the 'w' row from `right` (with NaN for 'y'). The 'a' rows appear twice due to the one-to-many relationship. Outer joins are useful when you want to see all data from both DataFrames, regardless of whether there are matches.

In [169]:
left.merge(right, how="cross")

x_x  y x_y  u
0   a  1   a  1
1   a  1   w  2
2   a  1   a  3
3   b  2   a  1
4   b  2   w  2
5   b  2   a  3

Lastly, we perform a cross join using `left.merge(right, how='cross')`. A cross join, also known as a Cartesian product, combines every row from the `left` DataFrame with every row from the `right` DataFrame. This results in a DataFrame with the number of rows equal to the product of the number of rows in both input DataFrames. Note how the column names are modified to avoid conflicts: 'x' becomes 'x_x' for the left DataFrame and 'x_y' for the right DataFrame. Cross joins can be useful in specific scenarios but should be used cautiously as they can result in very large datasets.

That's all for today!